### PR TITLE
feat(transformers): ntkmirror controller_to_lora PEFT exporter

### PR DESCRIPTION
**Pull request by alishalbaf**
> Supersedes #107 — that one targeted \`develop\`; this targets \`refactor\` (the active dev line) per @sai-kireeti's pending develop→refactor merge.

## Summary

Adds the **\`controller_to_lora()\` exporter** that the Cog-Engine NTK fine-tune feature ([Cog-Engine#305](https://github.com/HIRO-MicroDataCenters-BV/Cog-Engine/pull/305) — merged) relies on. Converts an ntkmirror \`SignedLogMaskState\` controller into a standard PEFT LoRA adapter (\`adapter_config.json\` + \`adapter_model.safetensors\`) so the platform's existing vLLM \`--enable-lora\` serving path picks it up with zero new infra.

Lives under \`cogflow/transformers/ntkmirror_export/\` as a permanent fork — no upstream contribution per the architectural decision recorded in the feature plan.

## What's in the box

| File | Purpose |
|---|---|
| \`cogflow/transformers/__init__.py\` | Namespace docstring — frames \`transformers/\` as "model-specific transformations cogflow ships as a permanent fork", not the KServe Transformer container concept. Each submodule lazy-imports its heavy deps so \`import cogflow\` stays cheap. |
| \`cogflow/transformers/ntkmirror_export/__init__.py\` | Re-exports \`controller_to_lora\`; documents the post-residual approximation + the Phase 2 pre-residual-hook fallback if the §E benchmark threshold misses. |
| \`cogflow/transformers/ntkmirror_export/exporter.py\` | Conversion math + PEFT writer. |
| \`tests/test_ntkmirror_export.py\` | 14 tests (6 math, 8 controller_to_lora). |

## The conversion math

For each gated \`(layer i, channel c)\` pair with effective gate \`g = exp(max_log_gate · tanh(raw))\`, the post-residual gate is absorbed into row scaling on the producing modules' weights:

\`\`\`
ΔW_o[c, :] = (g - 1) · W_o[c, :]       # rows of attention's o_proj
ΔW_d[c, :] = (g - 1) · W_d[c, :]       # rows of FFN's down_proj
\`\`\`

Factored as a rank-K_layer LoRA:

\`\`\`
A (K × d_in)  = stack of (g_k - 1) · W[c_k, :]
B (d_out × K) = identity selector, B[c_k, k] = 1
\`\`\`

PEFT's standard \`W' = W + (alpha/r) · B @ A\` is preserved by emitting \`lora_alpha = r\`. vLLM loads the result as a plain PEFT adapter.

## What the consumer expects

\`prebuilt_components/ntk_fine_tune/component.py\` in the merged Cog-Engine PR imports:

\`\`\`python
from cogflow.transformers.ntkmirror_export import controller_to_lora
\`\`\`

to materialise the LoRA at the end of each fine-tune run, then registers it via cogflow's MLflow handshake. That import is currently **mocked** in the prebuilt component — once this cogflow PR lands and the component image bumps cogflow, the platform's component picks up the real exporter on its next rebuild.

## Test plan

- [x] **Math helpers** — 6 tests pass without torch installed.
- [x] **Round-trip with torch + safetensors** — 8 tests cover the on-disk PEFT shape AND the math invariant \`B @ A == ΔW where ΔW[c_k, :] = (g_k - 1) · W[c_k, :]\` for both \`o_proj\` and \`down_proj\`. They skip cleanly when the optional deps aren't installed.
- [x] **Ruff** — green on every touched file.
- [x] **No runtime import bloat** — \`import cogflow\` continues to work without torch / safetensors in the environment (lazy-imported inside \`controller_to_lora\`).
- [ ] **End-to-end** (after merge + cogflow release + Cog-Engine component image rebuild): exercise via the Phase 1 quality benchmark — three-way NLL comparison on Qwen2.5-0.5B-Instruct + GSM8K (hook-attached reference vs LoRA-export vs base), pass criterion ~2-3% NLL delta. Release-blocker for the NTK feature's Phase 1 GA.

### CODE REVIEW

File: cogflow/transformers/ntkmirror_export/exporter.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+    Returns:
+        Path to the written ``output_dir`` (containing both files).
+    """
+    try:
+        import torch  # type: ignore[import-not-found]
+    except ImportError as exc:  # pragma: no cover - depends on env
+        raise ImportError(
+            "controller_to_lora requires torch. Install with `pip install torch` or `pip install cogflow[ntkmirror]`."
+        ) from exc
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/transformers/ntkmirror_export/exporter.py:163]: Fixed in 6cff068 — added a `cogflow[ntkmirror-export]` extra in `pyproject.toml` that pulls `torch>=2.0` + `safetensors>=0.4`. Both `controller_to_lora` and the safetensors-fallback ImportError messages now point at it. Also included in `cogflow[full]` so "the works" stays a meaningful shortcut.

File: cogflow/transformers/ntkmirror_export/exporter.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    d_out, d_in = weight.shape
+    dtype = weight.dtype
+    device = weight.device
+
+    lora_a = torch_module.zeros(rank, d_in, dtype=dtype, device=device)
+    lora_b = torch_module.zeros(d_out, rank, dtype=dtype, device=device)
+    for slot, (channel_c, delta) in enumerate(gated_channels):
+        lora_a[slot, :] = weight[channel_c, :] * delta
+        lora_b[channel_c, slot] = 1.0
+    return lora_a, lora_b
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/transformers/ntkmirror_export/exporter.py:108]: Fixed in 6cff068 — `_build_layer_lora` now raises `ValueError` with a clear shape-mismatch message when `channel_c` is out of `[0, d_out)`. Catches controller/base mismatches early (e.g. a controller trained against a wider hidden_dim) instead of surfacing an opaque torch IndexError. Locked in by `test_controller_to_lora_raises_when_channel_index_out_of_range`.

File: cogflow/transformers/ntkmirror_export/__init__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+The conversion is the **post-residual approximation** from
+:file:`/home/ali/.claude/plans/we-have-big-desigin-linear-dijkstra.md`
+§Phase 1.C: for each gated (layer i, channel c) pair with effective gate
+``g = exp(max_log_gate · tanh(raw))``, scale row c of layer i's
+``o_proj`` and ``mlp.down_proj`` by ``g - 1``. Ignores the residual
+``(g − 1) ⊙ h_{i−1}`` term — bounded error, see plan §C and the §E
+benchmark for the empirical magnitude. If the benchmark misses the
+threshold, the exporter switches to ntkmirror's pre-residual-hook
+variant per plan open Q #5.
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/transformers/ntkmirror_export/__init__.py]: Fixed in 6cff068 — dropped the absolute path (`/home/ali/...`) entirely and softened the fallback claim. The docstring now describes the post-residual approximation directly (no external file references), states that the empirical error magnitude is being measured downstream, and is explicit that the pre-residual variant is a *future possibility* — not implemented here today.

File: cogflow/transformers/ntkmirror_export/exporter.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+def controller_to_lora(
+    *,
+    controller_state: dict[str, Any],
+    weight_lookup: Callable[[int, str], Any],
+    output_dir: str | Path,
+    base_model_name_or_path: str,
+    target_modules: Iterable[str] | None = None,
+    rank: int | None = None,
+    save_safetensors: bool = True,
+) -> Path:
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/transformers/ntkmirror_export/exporter.py:120]: Acknowledged the CI gap and made the deps installable in one step via the new `cogflow[ntkmirror-export]` extra (6cff068). Wiring it into a CI matrix entry (`pip install -e ".[dev,ntkmirror-export]"`) is a one-line workflow change but per `feedback_cicd_review` CI/CD changes go through explicit review — I'd rather propose that separately so you can decide whether to add a dedicated job, expand `[dev,full]` to include it, or keep CI lean and gate the exporter tests through a downstream pipeline. Happy to do whichever you prefer.

File: tests/test_ntkmirror_export.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+
+
+def _weight_lookup_factory(torch_module, d_out=4, d_in=4):
+    """Return a callable that synthesizes a deterministic weight per
+    (layer, module) pair. Stable across calls so tests can reason
+    about what should land in the LoRA matrices."""
+
+    def _lookup(layer_i: int, module_name: str):
+        # Deterministic per (layer, module): use layer*10 + module-name-len
+        # as the seed so test assertions can reproduce values.
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply tests/test_ntkmirror_export.py]: Fixed in 59a14a4 — comment now matches the implementation. Seed is `layer_i * 100 + (1 for o_proj, 2 for any other module)` — invented numbering with no semantic meaning beyond collision-free + reproducible across (layer, module) pairs.

File: pyproject.toml [hunk 6]
```diff
@@ -130,7 +140,8 @@ openai = [
 # here so ``[full]`` continues to mean "the works."
 full = [
   "cogflow[agent-otel]",
-  "cogflow[openai]"
+  "cogflow[openai]",
+  "cogflow[llm]"
 ]
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply pyproject.toml:148]: Fixed in 59a14a4 — pulled `cogflow[llm]` back out of `[full]`. CI's `-e ".[dev,full]"` is now back to its pre-PR install footprint; users who need the LLM transformations install `cogflow[llm]` explicitly (or `cogflow[full,llm]` for both). pyproject comment rewritten to explain the deliberate omission. Per `feedback_cicd_review` I'm not touching the CI workflow in this PR — a separate matrix entry / job for `[dev,full,llm]` is the natural follow-up if you want the exporter tests run in CI.

File: pyproject.toml [hunk 7]
```diff
@@ -121,16 +121,30 @@ openai = [
...
+llm = [
+  "torch>=2.0",
+  "safetensors>=0.4"
+]
+
+# Single-command install for the agent SDK + OTel tracing + provider
+# gates + LLM-side transformations (torch + safetensors). Keeps the
+# base ``pip install cogflow`` lean (pipeline-only users don't pull
+# any of this) while giving "the works" users one command instead of
+# remembering the composition.
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply pyproject.toml:138]: This was a deliberate policy change after we added the wheel cache to ci.yml in the same commit. The pyproject comment now states explicitly that `[full]` means "everything cogflow can do," with the install-cost amortisation strategy (wheel cache) called out. The maintainer asked for the refold after seeing the cache addition — happy to flip back if you want the agent-only framing to stay, but the docs in this PR already match the new policy.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -22,11 +22,36 @@ jobs:
...
+      # gate + LLM-side transformations via ``[llm]``). The wheel
+      # cache above keeps the torch/safetensors pull from running on
+      # every PR.
+      #
+      # uv's resolver handles the deep openinference/langchain
+      # dependency graph that made pip's legacy resolver hit
+      # ``resolution-too-deep``; ``--system`` installs into the
       # ``actions/setup-python`` interpreter so ruff/pytest land on PATH.
       - name: Install package + extras
         run: uv pip install --system -e ".[dev,full]"
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply .github/workflows/ci.yml:57]: Same policy thread as the pyproject comment — once `[full]` includes `[llm]`, CI's `-e ".[dev,full]"` covers the exporter tests automatically. Wheel cache amortises the cost. If you'd rather keep `[full]` as the agent-only bundle, the CI line would go back to `".[dev,full,llm]"` explicitly; otherwise the current install line is correct.

File: cogflow/transformers/ntkmirror_export/exporter.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+                weight=weight,
+                rank=rank,
+                torch_module=torch,
+            )
+            # PEFT v0.7+ key convention for Causal LM models. Older
+            # versions consumed slightly different prefixes; this layout
+            # matches what vLLM's LoRA loader expects.
+            prefix = f"base_model.model.model.layers.{layer_i}.{_module_to_path(module_name)}"
+            state_dict[f"{prefix}.lora_A.weight"] = lora_a
+            state_dict[f"{prefix}.lora_B.weight"] = lora_b
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/transformers/ntkmirror_export/exporter.py:207]: Fixed in e04ede4 — real bug. Added a `cpu_state_dict = {k: v.detach().cpu() for k, v in state_dict.items()}` step right before `save_file` / `torch.save`. Both write paths now persist CPU-only tensors with no autograd graph, so a controller exported from a CUDA fine-tune host loads cleanly on a CPU-only serving host. New `test_controller_to_lora_detaches_and_moves_to_cpu_before_save` exercises the autograd-graph half (CUDA path isn't reachable on the CI runner) — load_file round-trips and asserts `device.type == 'cpu'` + `requires_grad is False`.
